### PR TITLE
Http auth implemented

### LINKED ISSUE #27439: Prompt user for authentication when HTTP request needs it
https://github.com/servo/servo/blob/08366323ef4ac77aa812d44499037bab4bc60eeb/components/net/http_loader.rs#L1571-L1578 and https://github.com/servo/servo/blob/08366323ef4ac77aa812d44499037bab4bc60eeb/components/net/http_loader.rs#L1606-L1612 require integration with the embedding layer in order to request user/password information from the browser UI.

We will need to store an [EmbedderProxy](https://github.com/servo/servo/blob/08366323ef4ac77aa812d44499037bab4bc60eeb/components/net/resource_thread.rs#L597) in [HttpState](https://github.com/servo/servo/blob/08366323ef4ac77aa812d44499037bab4bc60eeb/components/net/http_loader.rs#L83). This will allow us to send messages to the embedder from within the HTTP fetching code linked earlier.

We can then send two [EmbedderMsg::Prompt messages](https://doc.servo.org/embedder_traits/enum.PromptDefinition.html) and wait on the responses from the channel that is sent as part of that message, which will contain the username and password values chosen by the user.

To test this, run the following code under python3:
<details>

```
import http.server
import cgi
import base64
import json
from urllib.parse import urlparse, parse_qs


class CustomServerHandler(http.server.BaseHTTPRequestHandler):

    def do_HEAD(self):
        self.send_response(200)
        self.send_header('Content-type', 'application/json')
        self.end_headers()

    def do_AUTHHEAD(self):
        self.send_response(401)
        self.send_header(
            'WWW-Authenticate', 'Basic realm="Demo Realm"')
        self.send_header('Content-type', 'application/json')
        self.end_headers()

    def do_GET(self):
        key = self.server.get_auth_key()

        ''' Present frontpage with user authentication. '''
        if self.headers.get('Authorization') == None:
            self.do_AUTHHEAD()

            response = {
                'success': False,
                'error': 'No auth header received'
            }

            self.wfile.write(bytes(json.dumps(response), 'utf-8'))

        elif self.headers.get('Authorization') == 'Basic ' + str(key):
            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.end_headers()

            getvars = self._parse_GET()

            response = {
                'path': self.path,
                'get_vars': str(getvars)
            }

            base_path = urlparse(self.path).path
            if base_path == '/path1':
                # Do some work
                pass
            elif base_path == '/path2':
                # Do some work
                pass

            self.wfile.write(bytes(json.dumps(response), 'utf-8'))
        else:
            self.do_AUTHHEAD()

            response = {
                'success': False,
                'error': 'Invalid credentials'
            }

            self.wfile.write(bytes(json.dumps(response), 'utf-8'))

    def do_POST(self):
        key = self.server.get_auth_key()

        ''' Present frontpage with user authentication. '''
        if self.headers.get('Authorization') == None:
            self.do_AUTHHEAD()

            response = {
                'success': False,
                'error': 'No auth header received'
            }

            self.wfile.write(bytes(json.dumps(response), 'utf-8'))

        elif self.headers.get('Authorization') == 'Basic ' + str(key):
            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.end_headers()

            postvars = self._parse_POST()
            getvars = self._parse_GET()

            response = {
                'path': self.path,
                'get_vars': str(getvars),
                'get_vars': str(postvars)
            }

            base_path = urlparse(self.path).path
            if base_path == '/path1':
                # Do some work
                pass
            elif base_path == '/path2':
                # Do some work
                pass

            self.wfile.write(bytes(json.dumps(response), 'utf-8'))
        else:
            self.do_AUTHHEAD()

            response = {
                'success': False,
                'error': 'Invalid credentials'
            }

            self.wfile.write(bytes(json.dumps(response), 'utf-8'))

        response = {
            'path': self.path,
            'get_vars': str(getvars),
            'get_vars': str(postvars)
        }

        self.wfile.write(bytes(json.dumps(response), 'utf-8'))

    def _parse_POST(self):
        ctype, pdict = cgi.parse_header(self.headers.getheader('content-type'))
        if ctype == 'multipart/form-data':
            postvars = cgi.parse_multipart(self.rfile, pdict)
        elif ctype == 'application/x-www-form-urlencoded':
            length = int(self.headers.getheader('content-length'))
            postvars = cgi.parse_qs(
                self.rfile.read(length), keep_blank_values=1)
        else:
            postvars = {}

        return postvars

    def _parse_GET(self):
        getvars = parse_qs(urlparse(self.path).query)

        return getvars


class CustomHTTPServer(http.server.HTTPServer):
    key = ''

    def __init__(self, address, handlerClass=CustomServerHandler):
        super().__init__(address, handlerClass)

    def set_auth(self, username, password):
        self.key = base64.b64encode(
            bytes('%s:%s' % (username, password), 'utf-8')).decode('ascii')

    def get_auth_key(self):
        return self.key


if __name__ == '__main__':
    server = CustomHTTPServer(('', 8888))
    server.set_auth('demo', 'demo')
    server.serve_forever()
```

</details>
Then load `http://localhost:8888` in Servo (and any other browser to verify). The username `demo` with password `demo` should successfully pass the authentication check; everything else should continue to show the authentication prompt.

### PR DESCRIPTION
**Pull request by mustafapc19**
---

- [ X] `./mach build -d` does not report any errors
- [ X] `./mach test-tidy` does not report any errors
- [ X] These changes fix #27439 (GitHub issue number if applicable)

- [ ] There are tests for these changes OR
- [ X] These changes do not require tests because user input is required

### CODE REVIEW

**Comment by mustafapc19** [comment]
@jdm

File: components/net/http_loader.rs [hunk 1]
```diff
@@ -1603,17 +1634,20 @@ fn http_network_or_cache_fetch(
...
         // Step 4
-        // return http_network_or_cache_fetch(request, authentication_fetch_flag,
-        //                                    cors_flag, done_chan, context);
+        return http_network_or_cache_fetch(
+            http_request,
+            authentication_fetch_flag,
+            cors_flag,
+            done_chan,
+            context,
+        );
```
Comment (jdm) [inline components/net/http_loader.rs:1650]: These changes aren't quite right - [this step](https://fetch.spec.whatwg.org/#ref-for-proxy-authentication-entry%E2%91%A0) of the specification says to store the credentials somewhere and use them in [this step](https://fetch.spec.whatwg.org/#ref-for-proxy-authentication-entry), which is different than setting the Authorization header on the request. We should keep this code unmodified and add a FIXME comment saying that we can use the new prompt_user_for_credentials code but we need to store the result somewhere.

File: components/net/http_loader.rs [hunk 2]
```diff
@@ -1417,6 +1420,35 @@ fn http_network_or_cache_fetch(
         *done_chan = None;
     }
 
+    fn prompt_user_for_credentials(
+        context: &FetchContext,
+    ) -> std::option::Option<std::string::String> {
```
Comment (jdm) [inline components/net/http_loader.rs:1425]: This can be `Option<String>` instead.

File: components/net/http_loader.rs [hunk 3]
```diff
@@ -1417,6 +1420,35 @@ fn http_network_or_cache_fetch(
...
+        let local_embedder_proxy: EmbedderProxy;
+        {
+            local_embedder_proxy = context.state.embedder_proxy.lock().unwrap().clone();
+        }
```
Comment (jdm) [inline components/net/http_loader.rs:1429]: We can do `let embedder_proxy: EmbedderProxy = context.state.embedder_proxy.lock().unwrap().lock().clone();` instead here.

File: components/net/http_loader.rs [hunk 4]
```diff
@@ -1417,6 +1420,35 @@ fn http_network_or_cache_fetch(
...
+        let pass_prompt =
+            PromptDefinition::Input("Password".to_string(), "".to_string(), pass_sender);
+        let pass_msg = (
+            None,
+            EmbedderMsg::Prompt(pass_prompt, PromptOrigin::Trusted),
+        );
+        local_embedder_proxy.send(pass_msg);
+        let pass_str = pass_receiver.recv().unwrap()?;
+
+        serde::export::Some(base64::encode(&format!("{}:{}", name_str, pass_str)))
```
Comment (jdm) [inline components/net/http_loader.rs:1449]: This can be `Some(base64::encode(&format!("{}:{}", name_str, pass_str)))` instead.

**Review by jdm — CHANGES_REQUESTED** [review]
Looking good! We need to fix the authentication unit tests to not panic, too.

**Comment by mustafapc19** [comment]
Sure. I am just a little busy. I'll sure do it tho. I came this far

**Comment by jdm** [comment]
Yep! No rush, I just felt bad about leaving this PR without any feedback so far.

**Comment by mrobinson** [comment]
Closing in favor of #34620.

**Pull request closed without merging**